src/quest/manager.py

```diff
@@ -86,28 +94,47 @@ def _remove_workflow(self, workflow_id: str):
         self._workflow_data.remove(data)
 
     def _start_workflow(self,
-                        workflow_type: str, workflow_id: str, workflow_args, workflow_kwargs,
-                        background=False):
+                        workflow_type: str, workflow_id: str, workflow_args, workflow_kwargs, background=False):
         workflow_function = self._create_workflow(workflow_type)
 
+        future = asyncio.Future()
+        self._futures[workflow_id] = future
+
         workflow_manager.set(self)
 
         history = self._create_history(workflow_id)
         historian: Historian = Historian(workflow_id, workflow_function, history, serializer=self._serializer)
         self._workflows[workflow_id] = historian
 
         self._workflow_tasks[workflow_id] = (task := historian.run(*workflow_args, **workflow_kwargs))
-        if background:
-            task.add_done_callback(lambda t: self._remove_workflow(workflow_id))
+
+        if not background:
+            task.add_done_callback(lambda t: self._store_result(workflow_id, t))
+
+        task.add_done_callback(lambda t: self._remove_workflow(workflow_id))
+
+    def _store_result(self, workflow_id: str, task: asyncio.Task):
+        """Store the result or exception of a completed workflow"""
+        future = self._futures[workflow_id]
+
+        try:
+            result = task.result()
+            self._results[workflow_id] = result
+            future.set_result(result)
+        except Exception as e:
+            self._results[workflow_id] = e
```

Comment:
and `Exception` won't store to JSON. You'll need to store the module and name (like we do with step handling) and rehydrate the exception.

We should reuse the logic in step handling to do this. Maybe that functionality needs to be refactor out into a separate utility we can use in both places.